### PR TITLE
Eip1559 updates

### PR DESCRIPTION
**Pull request by i-norden**
I've updated the EIP1559 doc to reflect the implementation found [here](https://github.com/vulcanize/go-ethereum-EIP1559/tree/rebase) or [here](https://github.com/vulcanize/go-ethereum/tree/eip1559_rebase) if you want to see the tests passing. The PR for the implementation is now open [here](https://github.com/ethereum/go-ethereum/pull/20618).

The magicians thread for this can be found [here](https://ethereum-magicians.org/t/eip-1559-go-etheruem-implementation/3918). Please let me know if there is any additional feedback or updates to incorporate!

Thanks!

### CODE REVIEW

File: EIPS/eip-1559.md [hunk 1]
```diff
@@ -32,34 +34,61 @@ Ethereum currently prices transaction fees using a simple auction mechanism, whe
...
 <!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
 **Parameters**
-* `FORK_BLKNUM`: TBD
+* `INITIAL_FORK_BLKNUM`: TBD
 * `BASEFEE_MAX_CHANGE_DENOMINATOR`: 8
-* `SLACK_COEFFICIENT`: 3
-* `TARGET_GASUSED`: 8,000,000
+* `TARGET_GAS_USED`: 10,000,000
+* `MAX_GAS_EIP1559`: 16,000,000
+* `FINAL_FORK_BLKNUM`:  `INITIAL_FORK_BLKNUM + (MAX_GAS_EIP1559 / (10 * SLACK_COEFFICIENT))`
```
Comment (vbuterin) [inline EIPS/eip-1559.md]: Slack coefficient was removed, but is still used here.

Also, why a slack coefficient of only 1.6?

**Comment by eip-automerger** [comment]
Hi! I'm a bot, and I wanted to automerge your PR, but couldn't because of the following issue(s):

 - EIP 1559 requires approval from one of (@vbuterin, @econoar)

File: EIPS/eip-1559.md [hunk 2]
```diff
@@ -32,34 +34,61 @@ Ethereum currently prices transaction fees using a simple auction mechanism, whe
...
+
+For the gas limit:
+
+* `MAX_GAS_EIP1559` acts as the hard in-protocol gas limit, instead of the gas limit calculated using the previously existing formulas
+* The `GASLIMIT` field in the block header is the gas limit for the EIP1559 gas pool, and over the transition period this value increases until it reaches `MAX_GAS_EIP1559` at `FINAL_FORK_BLKNUM`
+* The gas limit for the legacy gas pool is `MAX_GAS_EIP1559 - GASLIMIT`, as `GASLIMIT` increases towards `MAX_GAS_EIP1559` gas is moved from the legacy pool into the EIP1559 pool until all of the gas is in the EIP1559 pool
+* At `block.number == INITIAL_FORK_BLKNUM`, let `GASLIMIT = (MAX_GAS_EIP1559 / 2)` so that the gas maximum is split evenly between the legacy and EIP1559 gas pools
+* As `block.number` increases towards `FINAL_FORK_BLKNUM`, at every block we shift `EIP1559_GAS_INCREMENT_AMOUNT` from the legacy pool into the EIP1559 gas pool
+* At `block.number >= FINAL_FORK_BLKNUM` the entire `MAX_GAS_EIP1559` is assigned to the EIP1559 gas pool and the legacy pool is empty
+* We enforce a per-transaction gas limit of 8,000,000.``
```
Comment (vbuterin) [inline EIPS/eip-1559.md]: 8,000,000 should be a parameter, no?

File: EIPS/eip-1559.md [hunk 3]
```diff
@@ -32,34 +34,61 @@ Ethereum currently prices transaction fees using a simple auction mechanism, whe
...
+
+For the gas price:
+
+* We add a new field to the block header, `BASEFEE`
+  * `BASEFEE` is maintained under consensus by the ethash engine
+* At `block.number == INITIAL_FORK_BLKNUM` we set `BASEFEE = INITIAL_BASEFEE`
+* As a default strategy, miners set `BASEFEE` as follows.
+  * Let `delta = block.gas_used - TARGET_GASUSED` (possibly negative).
+  * Set `BASEFEE = PARENT_BASEFEE + PARENT_BASEFEE * delta // TARGET_GASUSED // BASEFEE_MAX_CHANGE_DENOMINATOR`
+  * Clamp the resulting `BASEFEE` inside of the allowable bounds if needed, where a valid `BASEFEE` is one such that `abs(BASEFEE - PARENT_BASEFEE) <= max(1, PARENT_BASEFEE // BASEFEE_MAX_CHANGE_DENOMINATOR)`
```
Comment (vbuterin) [inline EIPS/eip-1559.md]: Are we letting miners adjust the basefee? I thought we are forcing the specific change in the protocol?

File: EIPS/eip-1559.md (hunk 2 above)
Comment (i-norden) [reply EIPS/eip-1559.md]: Yes, thanks! It is a parameter in the implementation, but I didn't reflect that here. Will update. Although it looks like there is some more discussion about what the per-transaction gas limit should in the magicians thread: https://ethereum-magicians.org/t/eip-1559-go-ethereum-implementation/3918/15?u=i-norden

File: EIPS/eip-1559.md (hunk 3 above)
Comment (i-norden) [reply EIPS/eip-1559.md]: I haven't had the time to thoroughly model the BASEFEE function, but in my limited experimentation there were scenarios where the specific change enforced by the protocol would result in a BASEFEE that falls out of the bounds described above. Only in that case is the result clamped within those bounds.

I will update all the references to "miners adjusting/setting BASEFEE" to reflect it being the protocol. I think this is a bit of semantic misunderstanding on my part, in my eyes the miners are still "setting" the BASEFEE in the blocks they mine, even if they have no control over what that set value is.

File: EIPS/eip-1559.md [hunk 4]
```diff
@@ -17,11 +17,13 @@ The current "first price auction" fee model in Ethereum is inefficient and needl
...
 
-* A "premium" gasprice which gets added onto the BASEFEE gasprice, which can either be set to a fairly low value (eg. 1 gwei) to compensate miners for uncle rate risk or to a high value to compete during sudden bursts of activity. The BASEFEE gets burned, the premium is given to the miner.
+* A gas premium which gets added onto the base fee to calculate the gas price. The gas premium can either be set to a fairly low value (eg. 1 gwei) to compensate miners for uncle rate risk or to a high value to compete during sudden bursts of activity. The base fee gets burned, the gas premium is given to the miner.
 
-* A "cap" which represents the maximum total that the transaction sender would be willing to pay to get included.
+* A fee cap which represents the maximum total (base fee + gas premium) that the transaction sender would be willing to pay to get their transaction included.
+
+The current miner-voting based gas limit is changed to a hard-coded gas limit of 16 million. Instead of miners directly adjusting the gas limit in response to changes in network demand they adjust the base fee to apply economic pressure towards a target gas usage of 10 million.
```
Comment (edmundedgar) [inline EIPS/eip-1559.md]: This reads like miners still have control of the parameter. If I'm right in thinking they don't it would be clearer to say "the protocol adjusts the gas limit" or "the gas limit is adjusted".

Comment (i-norden) [reply EIPS/eip-1559.md]: I agree, thanks for pointing this out! I've updated it to hopefully be more clear.

**Comment by Souptacular** [comment]
Just wanted to pop in and say great work!

File: EIPS/eip-1559.md (hunk 1 above)
Comment (ksheni56) [reply EIPS/eip-1559.md]: gogo I want to see ethereum at #1

Comment (i-norden) [reply EIPS/eip-1559.md]: Sorry I missed this comment when it was made.

I've removed any mention of the slack coefficient from the doc, as it had already been removed from the implementation.

The slack coefficient was never being used for anything other than as the scalar that described the ratio between max gas used and target gas used. It was 2 when target gas used was 8m and max gas used was 16m, but after switching to a target gas used of 10m it became 1.6.

File: EIPS/eip-1559.md (hunk 3 above)
Comment (vbuterin) [reply EIPS/eip-1559.md]: Maybe to avoid simplicity reword this as:

* Let `gas_delta = block.gas_used - TARGET_GASUSED` (possibly negative).
* Let `basefee_delta = PARENT_BASEFEE * delta // TARGET_GASUSED // BASEFEE_MAX_CHANGE_DENOMINATOR`
* Let `max_basefee_delta = max(1, PARENT_BASEFEE // BASEFEE_MAX_CHANGE_DENOMINATOR)`
* Set `BASEFEE = PARENT_BASEFEE + max(-max_basefee_delta, min(max_basefee_delta, basefee_delta))`

Comment (vbuterin) [reply EIPS/eip-1559.md]: Actually, it's also important to specify cleanly what we mean by // in the negative case. Is `-7 // 5` equal to `-1` or `-2`? In python it's the latter, in some languages it's the former. Perhaps avoid negative numbers entirely?

* Let `max_basefee_delta = max(1, PARENT_BASEFEE // BASEFEE_MAX_CHANGE_DENOMINATOR)`
* If `block.gas_used >= TARGET_GASUSED`:
    * Let `gas_delta = block.gas_used - TARGET_GASUSED`
    * Let `basefee_delta = PARENT_BASEFEE * gas_delta // TARGET_GASUSED // BASEFEE_MAX_CHANGE_DENOMINATOR`
    * Set `BASEFEE = PARENT_BASEFEE + min(max_basefee_delta, basefee_delta)`
* `block.gas_used < TARGET_GASUSED`:
    * Let `gas_delta = TARGET_GASUSED - block.gas_used`
    * Let `basefee_delta = PARENT_BASEFEE * gas_delta // TARGET_GASUSED // BASEFEE_MAX_CHANGE_DENOMINATOR`
    * Set `BASEFEE = max(0, PARENT_BASEFEE - min(max_basefee_delta, basefee_delta))`

Comment (i-norden) [reply EIPS/eip-1559.md]: That's a great point, we are using `big.Int` division which is true Euclidean division where the remainder is bound above 0 so we round to negative infinity and would get `-2`. We can make that the specification, or avoid negative numbers entirely. What do you think is best?
